File: libraries/Microsoft.Bot.Builder.Expressions/Constant.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.Bot.Builder.Expressions
{
    /// <summary>
    /// Constant expression.
    /// </summary>
    public class Constant : Expression
    {
        private object _value;

        /// <summary>
        /// Initializes a new instance of the <see cref="Constant"/> class.
        /// Construct an expression constant.
        /// </summary>
        /// <param name="value">Constant value.</param>
        public Constant(object value = null)
            : base(new ExpressionEvaluator(ExpressionType.Constant, (expression, state) => ((expression as Constant).Value, null)))
        {
            Value = value;
        }

        /// <summary>
        /// Gets or sets constant value.
        /// </summary>
        /// <value>
        /// Constant value.
        /// </value>
        public object Value
        {
            get
            {
                return _value;
            }

            set
            {
                Evaluator.ReturnType =
                      value is string ? ReturnType.String
                      : value.IsNumber() ? ReturnType.Number
                      : value is bool ? ReturnType.Boolean
                      : ReturnType.Object;
                _value = value;
            }
        }

        public override string ToString()
        {
            if (Value == null)
            {
                return "null";
            }
            else if (Value is string)
            {
                return $"'{Value}'";
            }
            else
            {
                return Value?.ToString();
            }
        }
    }
}

File: libraries/AdaptiveExpressionsCpp/AdaptiveExpressionsCpp/Code/Constant.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "Constant.h"
#include "FunctionUtils.h"
#include "ExpressionType.h"

Constant::Constant(std::any value) :
    Expression(new ConstantExpressionEvaluator())
{
    setValue(value);
}

std::any Constant::getValue()
{
    return m_value;
}

EvaluateExpressionLambda Constant::Evaluator()
{
    return [](Expression* expression, void* state, void* options)
    {
        Constant* constant = (Constant*)expression;
        return ValueErrorTuple(constant->getValue(), std::string());
    };;
}

void Constant::setValue(std::any value)
{
    // Have to figure out how to check if it's an array
    m_evaluator->setReturnType(FunctionUtils::isOfType<std::string>(value) ? ReturnType::String :
        FunctionUtils::isNumber(value) ? ReturnType::Number :
        FunctionUtils::isOfType<bool>(value) ? ReturnType::String :
        // FunctionUtils::isOfType<std::vector<template T>> ? ReturnType::Array :
        ReturnType::Object);

    m_value = value;
}

ConstantExpressionEvaluator::ConstantExpressionEvaluator() :
    ExpressionEvaluator(ExpressionType::Constant, ReturnType::Object)
{
}

ValueErrorTuple ConstantExpressionEvaluator::TryEvaluate(Expression* expression, void* state, Options* options)
{
    Constant* constant = (Constant*)expression;
    return ValueErrorTuple(constant->getValue(), std::string());
}

void ConstantExpressionEvaluator::ValidateExpression(Expression* expression)
{
}
